pkg/build/controller/buildpod/factory.go

```diff
@@ -0,0 +1,213 @@
+package buildpod
+
+import (
+	"errors"
+	"time"
+
+	"github.com/golang/glog"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	kcontroller "k8s.io/kubernetes/pkg/controller"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildclient "github.com/openshift/origin/pkg/build/client"
+	"github.com/openshift/origin/pkg/build/controller/policy"
+	strategy "github.com/openshift/origin/pkg/build/controller/strategy"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	osclient "github.com/openshift/origin/pkg/client"
+)
+
+const (
+	// We must avoid processing build pods until the build and pod stores have synced.
+	// If they haven't synced, to avoid a hot loop, we'll wait this long between checks.
+	storeSyncedPollPeriod = 100 * time.Millisecond
+	maxRetries            = 60
```

Comment:
> I disagree that listening on all imagestreams in the cluster to see if the imagestream we need has appeared yet and then go back and figure out what build was waiting for that imagestream and prod it forward

fwiw, this works elsewhere w/o a problem. Also seems to be the case in https://github.com/openshift/origin/pull/13242

> and if we don't get the event

I think all of the events are delivered eventually but @smarterclayton knows the gory details of watch better.